gbpservice/neutron/services/servicechain/drivers/simplechain_driver.py

```diff
@@ -487,26 +487,37 @@ def delete(self, stack_id):
     def get(self, stack_id):
         return self.stacks.get(stack_id)
 
-    @property
-    def keystone(self):
-        if not self._keystone:
-            keystone_conf = cfg.CONF.keystone_authtoken
-            if keystone_conf.get('auth_uri'):
-                auth_url = keystone_conf.auth_uri
-                if not auth_url.endswith('/v2.0/'):
-                    auth_url += '/v2.0/'
-            else:
-                auth_url = ('%s://%s:%s/v2.0/' % (
-                    keystone_conf.auth_protocol,
-                    keystone_conf.auth_host,
-                    keystone_conf.auth_port))
-            user = (keystone_conf.get('admin_user') or keystone_conf.username)
-            pw = (keystone_conf.get('admin_password') or
-                  keystone_conf.password)
-            self._keystone = keyclient.Client(
-                username=user, password=pw, auth_url=auth_url,
-                tenant_id=self.tenant)
-        return self._keystone
-
     def _get_auth_token(self, tenant):
-        return self.keystone.get_token(tenant)
+        keystone_obj = get_keystone_client()
+        return keystone_obj.get_token(tenant)
+
+    def _get_admin_auth_token(self):
+        keystone_obj = get_keystone_client()
+        admin_tenant_obj = get_admin_tenant_object()
+        return keystone_obj.get_token(admin_tenant_obj.id)
+
+def get_keystone_client():
+    keystone_conf = cfg.CONF.keystone_authtoken
+    if keystone_conf.get('auth_uri'):
+        auth_url = keystone_conf.auth_uri
+        if not auth_url.endswith('/v2.0/'):
+            auth_url += '/v2.0/'
+    else:
+        auth_url = ('%s://%s:%s/v2.0/' % (
+            keystone_conf.auth_protocol,
+            keystone_conf.auth_host,
+            keystone_conf.auth_port))
+    user = (keystone_conf.get('admin_user') or keystone_conf.username)
+    pw = (keystone_conf.get('admin_password') or
+          keystone_conf.password)
+    tenant_name = keystone_conf.get('admin_tenant_name')
```

Comment:
Magesh, the code in Noiro branch *does not* use admin tenant. The Intercloud emo works with normal tenant:
https://github.com/noironetworks/group-based-policy/blob/noiro/gbpservice/neutron/services/servicechain/drivers/simplechain_driver.py#L501-L511